SofaKernel/modules/SofaSimpleFem/src/SofaSimpleFem/TriangleFEMUtils.inl

```diff
@@ -0,0 +1,284 @@
+    edgey.normalize();
+
+    r[0][0] = edgex[0];
+    r[0][1] = edgex[1];
+    r[0][2] = edgex[2];
+    r[1][0] = edgey[0];
+    r[1][1] = edgey[1];
+    r[1][2] = edgey[2];
+    r[2][0] = edgez[0];
+    r[2][1] = edgez[1];
+    r[2][2] = edgez[2];
+}
+
+
+// -------------------------------------------------------------------------------------------------------------
+// --- Compute displacement vector D as the difference between current current position 'p' and initial position
+// --- expressed in the co-rotational frame of reference
+// -------------------------------------------------------------------------------------------------------------
+template<class DataTypes>
+constexpr void TriangleFEMUtils<DataTypes>::computeDisplacementLarge(Displacement& D, const Transformation& R_0_2, const type::fixed_array<Coord, 3>& rotatedInitCoord, const Coord& pA, const Coord& pB, const Coord& pC)
+{
+    // positions of the deformed and displaced triangle in its local frame
+    Coord deforme_b = R_0_2 * (pB - pA);
+    Coord deforme_c = R_0_2 * (pC - pA);
+
+    // displacements in the local frame
+    D[0] = 0;
+    D[1] = 0;
+    D[2] = rotatedInitCoord[1][0] - deforme_b[0];
+    D[3] = 0;
+    D[4] = rotatedInitCoord[2][0] - deforme_c[0];
+    D[5] = rotatedInitCoord[2][1] - deforme_c[1];
+}
+
+
+// --------------------------------------------------------------------------------------
+// ---	Compute F = J * stress;
+// --------------------------------------------------------------------------------------
+template<class DataTypes>
+constexpr void TriangleFEMUtils<DataTypes>::computeForceLarge(Displacement& F, const StrainDisplacement& J, const type::Vec<3, Real>& stress)
+{
+    //	F = J * stress;
+    // Optimisations: The following values are 0 (per computeStrainDisplacement )
+    // \       0        1        2
+    // 0   J[0][0]      0      J[0][2]
+    // 1       0     J[1][1]   J[1][2]
+    // 2   J[2][0]      0      J[2][2]
+    // 3       0     J[3][1]   J[3][2]
+    // 4       0        0      J[4][2]
+    // 5       0     J[5][1]     0
+
+    F[0] = J[0][0] * stress[0] + J[0][2] * stress[2];
+    F[1] = J[1][1] * stress[1] + J[1][2] * stress[2];
+    F[2] = J[2][0] * stress[0] + J[2][2] * stress[2];
+    F[3] = J[3][1] * stress[1] + J[3][2] * stress[2];
+    F[4] = J[4][2] * stress[2];
+    F[5] = J[5][1] * stress[1];
+}
+
+
+// ------------------------------------------------------------------------------------------------------------
+// --- Compute the strain-displacement matrix where (pA, pB, pC) are the coordinates of the 3 nodes of a triangle
+// ------------------------------------------------------------------------------------------------------------
+template<class DataTypes>
+constexpr void TriangleFEMUtils<DataTypes>::computeStrainDisplacementGlobal(StrainDisplacement& J, const Coord& pA, const Coord& pB, const Coord& pC)
+{
+    Coord ab_cross_ac = cross(pB - pA, pC - pA);
+    Real determinant = ab_cross_ac.norm();
```

Comment:
determinant will always be non-zero ?

Reply:
could you check if my test can be optimized with new c++ tricks